python/ray/util/ray_executor.py

```diff
@@ -0,0 +1,237 @@
+        return future
+
+    def map(self, fn, *iterables, timeout=None, chunksize=1):
+        """Returns an iterator equivalent to `map(fn, iter)`.
+
+        Args:
+            fn: A callable that will take as many arguments as there are
+                passed iterables.
+            timeout: The maximum number of seconds to wait. If None, then there
+                is no limit on the wait time.
+            chunksize: The size of the chunks the iterable will be broken into
+                before being passed to a child process.
+
+        Returns:
+            An iterator equivalent to: `map(func, *iterables)` but the calls may
+            be evaluated out-of-order.
+
+        Raises:
+            TimeoutError: If the entire result iterator could not be generated
+                before the given timeout.
+            Exception: If `fn(*args)` raises for any values.
+
+        Usage example:
+
+        .. code-block:: python
+
+            with RayExecutor() as ex:
+                futures = ex.map(lambda x: x * x, [100, 100, 100])
+                results = [future.result() for future in futures()]
+
+        """
+        self._check_shutdown_lock()
+        results_list = []
+        for chunk in self._get_chunks(*iterables, chunksize=chunksize):
+            if self.__actor_pool:
+                results = self.__actor_pool.map(
+                    lambda a, v: a.actor_function.remote(partial(fn, *v)), chunk
+                )
+            else:
+                results = self._map(self.submit, fn, chunk, timeout)
+            results_list.append(results)
+        return itertools.chain(*results_list)
+
+    @staticmethod
+    def _get_chunks(*iterables, chunksize):
+        """
+        https://github.com/python/cpython/blob/main/Lib/concurrent/futures/process.py#L186
+        Iterates over zip()-ed iterables in chunks.
+        """
+        it = zip(*iterables)
+        while True:
+            chunk = tuple(itertools.islice(it, chunksize))
+            if not chunk:
+                return
+            yield chunk
+
+    @staticmethod
+    def _result_or_cancel(fut, timeout=None):
+        """
+        From concurrent.futures
+        """
+        try:
+            try:
+                return fut.result(timeout)
+            finally:
+                fut.cancel()
+        finally:
+            # Break a reference cycle with the exception in self._exception
+            del fut
+
+    def _map(self, submit_fn, fn, iterables, timeout=None):
+        """
+        This was adapted from concurrent.futures.Executor.map.
+        """
+        if timeout is not None:
+            end_time = timeout + time.monotonic()
+
+        fs = [submit_fn(fn, *args) for args in iterables]
+
+        # Yield must be hidden in closure so that the futures are submitted
+        # before the first iterator value is required.
+        def result_iterator():
+            try:
+                # reverse to keep finishing order
+                fs.reverse()
```

Comment:
Does this return the task that's finished first? E.g., 

[task A, task B, task C]

if task B is completed before A and C, wouldn't this block until C is over with this logic?

Reply:
This should return tasks (as futures) in the reverse order of their submission, not completion. Might be worth adding a test to verify, though.

Reply:
> reverse order of their submission

Is this the same semantic as other *PoolExecutor? Why do we iterate in a reversed order not the submission order?

Reply:
Yes, this is used by `Executor`. The reason is that `pop()` is used in generating the result iterator, so items are taken from the end of the list.

Reply:
i.e. this will retain submission order